crypto/rsa/rsa_gen.c

```diff
@@ -71,6 +71,188 @@ int RSA_generate_multi_prime_key(RSA *rsa, int bits, int primes,
     return rsa_keygen(rsa->libctx, rsa, bits, primes, e_value, cb, 0);
 }
 
+DEFINE_STACK_OF(BIGNUM)
+
+/*
+ * Given input values, q, p, n, d and e, derive the exponents 
+ * and coefficients for each prime in this key, placing the result
+ * on thier respective exps and coeffs stacks
+ */
+int rsa_multiprime_derive(RSA *rsa, int bits, int primes,
+                                 BIGNUM *e_value,
+                                 STACK_OF(BIGNUM) *factors, STACK_OF(BIGNUM) *exps,
+                                 STACK_OF(BIGNUM) *coeffs)
+{
+    STACK_OF(BIGNUM) *pplist = NULL, *pdlist = NULL;
+    BIGNUM *factor = NULL, *newpp = NULL, *newpd = NULL;
+    BIGNUM *dval = NULL, *newexp = NULL, *newcoeff = NULL;
+    BIGNUM *p = NULL, *q = NULL;
+    BIGNUM *r0 = NULL, *r1 = NULL, *r2 = NULL;
+    BN_CTX *ctx = NULL;
+    int i;
+    int ret = 0;
+
+
+    ctx = BN_CTX_new_ex(rsa->libctx);
+    if (!ctx)
+        goto err;
+
+    BN_CTX_start(ctx);
+
+    pplist = sk_BIGNUM_new_null();
+    if (!pplist)
+        goto err;
+
+    pdlist = sk_BIGNUM_new_null();
+    if (!pdlist)
+        goto err;
+
+    r0 = BN_new();
```

Comment:
You can use
````
r0 = BN_CTX_get(ctx);
r1 = BN_CTX_get(ctx);
r2 = BN_CTX_get(ctx);
if (r2 == NULL)
    goto err;
````

Reply:
Same for pr0 etc below.. which gets rid of the extra {}

Reply:
the pr0 calculation was a debug check I left in there, should be removed entirely, but I'll alter the others to use the context